cadence/contracts/GiefMe.cdc

```diff
@@ -0,0 +1,263 @@
+//  /'_ `\/\ \  /'__`\ \ ,__\      /' __` __`\  /'__`\
+// /\ \L\ \ \ \/\  __/\ \ \_/      /\ \/\ \/\ \/\  __/
+// \ \____ \ \_\ \____\\ \_\       \ \_\ \_\ \_\ \____\
+//  \/___L\ \/_/\/____/ \/_/        \/_/\/_/\/_/\/____/
+//    /\____/
+//    \_/__/
+// GiefMe - wrap NFT gifts in a box and send them to your friends.
+// The gifts can be claimed by passing the correct password.
+//
+pub contract GiefMe {    
+    /**//////////////////////////////////////////////////////////////
+    //                           PATHS                            //
+    /////////////////////////////////////////////////////////////**/
+
+    pub let GiefMeStoragePath: StoragePath
+    pub let GiefMePublicPath: PublicPath
+    pub let GiefMePrivatePath: PrivatePath
+
+    /**//////////////////////////////////////////////////////////////
+    //                            EVENTS                           //
+    /////////////////////////////////////////////////////////////**/
+
+    pub event Packed(gieft: UInt64, nfts: [UInt64])
+    pub event Added(gieft: UInt64, nft: UInt64, type: String, name: String, thumbnail: String)
+    pub event Removed(gieft: UInt64, nft: UInt64, type: String, name: String, thumbnail: String)
+    pub event Claimed(gieft: UInt64, nft: UInt64, type: String, name: String, thumbnail: String, gifter: Address?, giftee: Address?)
+
+    /**//////////////////////////////////////////////////////////////
+    //                         INTERFACES                          //
+    /////////////////////////////////////////////////////////////**/
+
+    /// Gieft
+
+    pub resource interface GieftPublic {
+        pub let password: [UInt8]
+        pub fun borrowClaimableNFT(): &NonFungibleToken.NFT?
+        pub fun claimNft(password: String, collection: &AnyResource{NonFungibleToken.CollectionPublic, MetadataViews.ResolverCollection})
+        pub fun getNftIDs(): [UInt64]
+    }
+
+    /// GieftCollection
+
+    pub resource interface GieftCollectionPublic {
+        pub fun borrowGieft(_ gieft: UInt64): &Gieft{GieftPublic}?
+        pub fun getGieftIDs(): [UInt64]
+    }
+
+    pub resource interface GieftCollectionPrivate {
+        pub fun packGieft(name: String, password: [UInt8], nfts: @{UInt64: NonFungibleToken.NFT}, registryCapability: Capability<&FindRegistry.Registry{FindRegistry.RegistryPublic, FindRegistry.RegistryPrivate}>?)
+        pub fun addNftToGieft(gieft: UInt64, nft: @NonFungibleToken.NFT)
+        pub fun unpackGieft(gieft: UInt64): @{UInt64: NonFungibleToken.NFT} 
+    }
+
+    /**//////////////////////////////////////////////////////////////
+    //                         RESOURCES                           //
+    /////////////////////////////////////////////////////////////**/
+
+    /// Gieft
+    /// A collection of NFTs that can be claimed by passing the correct password
+
+    pub resource Gieft: GieftPublic {
+        ///  The name of the gieft
+        pub let name: String
```

Comment:
can we add a { String : Anystruct} propertyBag here, just incase we need to add more state.